stix2/sources/__init__.py

```diff
@@ -136,194 +221,122 @@ def add(self, stix_objs):
-            data from the data source.
+        Implement: Similar to get() except returns list of all object versions of
+        the specified "id". In addition, implement the specific data
+        source API calls, processing, functionality required for retrieving
+        data from the data source.
 
         Args:
             stix_id (str): The id of the STIX 2.0 object to retrieve. Should
                 return a list of objects, all the versions of the object
                 specified by the "id".
 
-            _composite_filters (list): list of filters passed from the
-                Composite Data Source
+            _composite_filters (set): set of filters passed from the parent
+                CompositeDataSource, not user supplied
 
         Returns:
-            stix_objs (list): a list of STIX objects (where each object is a
-                STIX object)
+            stix_objs (list): a list of STIX objects
 
         """
         raise NotImplementedError()
 
     def query(self, query, _composite_filters=None):
         """
-        Fill:
-            -implement the specific data source API calls, processing,
-            functionality required for retrieving query from the data source
+        Implement:Implement the specific data source API calls, processing,
+        functionality required for retrieving query from the data source
 
         Args:
             query (list): a list of filters (which collectively are the query)
                 to conduct search on
 
-            _composite_filters (list): a list of filters passed from the
-                Composite Data Source
+            _composite_filters (set): a set of filters passed from the parent
+                CompositeDataSource, not user supplied
 
         Returns:
+            stix_objs (list): a list of STIX objects
 
         """
         raise NotImplementedError()
 
     def add_filters(self, filters):
-        """Add multiple filters to be applied to all queries for STIX objects.
-
-        Args:
-            filters (list): list of filters (dict) to add to the Data Source.
-
-        """
-        for filter in filters:
-            self.add_filter(filter)
-
-    def add_filter(self, filter):
         """Add a filter to be applied to all queries for STIX objects.
 
         Args:
-            filter: filter to add to the Data Source.
+            filters (list or Filter obj): filter(s) to add to the Data Source.
 
         """
-        # check filter field is a supported STIX 2.0 common field
-        if filter.field not in STIX_COMMON_FIELDS:
-            raise ValueError("Filter 'field' is not a STIX 2.0 common property. Currently only STIX object common properties supported")
-
-        # check filter operator is supported
-        if filter.op not in FILTER_OPS:
-            raise ValueError("Filter operation (from 'op' field) not supported")
+        if isinstance(filters, list) or isinstance(filters, set):
```

Comment:
Why get rid of the `add_filter`/`add_filters` distinction? It seems a bit circuitous to have a check every time through the function of whether it's a list or not. This approach does allow you to pass in a list-of-lists-of-filters, but I'm not sure we want/need that complexity.

Reply:
That was something I was going to bring up. I didn't like having the distinction, just based on other APIs I have used they have had one method to do functionality like this.  Im okay with it either way, not that critical.

To your latter point, yep noticed that. Obviously I dont think anyone would use the method that way. I just liked that you could supply a list, set or Filter to one method and it just handles it all. Alternatively, you could just make the method accept a list, or go back to two distinct methods.

Reply:
I like the idea of a single function that handles all cases, and that is what's been done in the `markings` modules. However, [this StackOverflow answer](https://softwareengineering.stackexchange.com/a/296765/) suggests consensus is to use two functions, the plural version calling the singular.

I don't care too much about this either, but regardless we should be consistent throughout the library as to whether we have both a singular and a plural function, or just a single function for both cases.